### PR TITLE
Add ability to manage content metadata attribute externally

### PR DESCRIPTION
**Pull request by stonko1994**
## Description
This will add the possibility to override content metadata attributes externally.

### CODE REVIEW

File: src/ts/ContentMetadataBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+interface CustomContentMetadata {
+  [key: string]: string;
+}
+
+export class ContentMetadataBuilder {
+
+  private readonly logger: Conviva.LoggingInterface;
+  private readonly contentMetadata: Conviva.ContentMetadata;
+
+  // intern metadata fields to enable merging / overriding
```
Comment (bitmovin-kenny) [inline src/ts/ContentMetadataBuilder.ts]:
```suggestion
  // internal metadata fields to enable merging / overriding
```

File: src/ts/ContentMetadataBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      this.logger.consoleLog(
+        '[ Conviva Analytics ] Playback has started. Only some metadata attributes will be updated',
+        Conviva.SystemSettings.LogLevel.WARNING,
+      );
+    }
+
+    this.metadataOverrides = newValue;
+  }
+
+  // Is needed to determine which attributes are allowed to update
```
Comment (bitmovin-kenny) [inline src/ts/ContentMetadataBuilder.ts]: this comment looks out of place

File: src/ts/ConvivaAnalytics.ts [hunk 3]
```diff
@@ -144,19 +123,18 @@ export class ConvivaAnalytics {
...
+    if (this.isValidSession()) {
+      this.logger.consoleLog('There is already a session running.', Conviva.SystemSettings.LogLevel.WARNING);
+      return;
     }
 
-    if (assetName) {
-      this.metadataOverrides.assetName = assetName;
+    // This could be called before source loaded.
+    // Without setting the asset name on the content metadata when initializing the session the SDK will throw errors.
+    if (!this.player.getSource() && !this.contentMetadataBuilder.assetName) {
```
Comment (bitmovin-kenny) [inline src/ts/ConvivaAnalytics.ts:137]: shouldn't this be part of `isValidSession` ?

File: src/ts/ConvivaAnalytics.ts [hunk 4]
```diff
@@ -144,19 +123,18 @@ export class ConvivaAnalytics {
...
+  public initializeSession(): void {
+    if (this.isValidSession()) {
+      this.logger.consoleLog('There is already a session running.', Conviva.SystemSettings.LogLevel.WARNING);
+      return;
     }
 
-    if (assetName) {
-      this.metadataOverrides.assetName = assetName;
+    // This could be called before source loaded.
+    // Without setting the asset name on the content metadata when initializing the session the SDK will throw errors.
```
Comment (bitmovin-kenny) [inline src/ts/ConvivaAnalytics.ts]: `the session the SDK will throw errors` something is wrong here

File: src/ts/ConvivaAnalytics.ts [hunk 5]
```diff
@@ -204,6 +182,31 @@ export class ConvivaAnalytics {
...
   }
 
+  /**
+   * Will update the contentMetadata which are tracked with conviva.
+   *
+   * If there is an active session only permitted values will be updated and propagated immediately.
+   * If there is no active session the values will set on session creation.
+   *
+   * Attributes set via this method will override automatic tracked once.
+   * @param metadata
```
Comment (bitmovin-kenny) [inline src/ts/ConvivaAnalytics.ts]: missing jsDoc

File: src/ts/ConvivaAnalytics.ts [hunk 6]
```diff
@@ -204,6 +182,31 @@ export class ConvivaAnalytics {
...
+   * @see ContentMetadataBuilder for more information about permitted attributes
+   */
+  public updateContentMetadata(metadata: Metadata) {
```
Comment (bitmovin-kenny) [inline src/ts/ConvivaAnalytics.ts]: i would suggest `metadataOverrides` as param name

**Review by bitmovin-kenny — CHANGES_REQUESTED** [review]
generally looking good, some nitpicks

File: src/ts/ContentMetadataBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    this.contentMetadata.encodedFrameRate = this.metadataOverrides.encodedFrameRate || this.metadata.encodedFrameRate;
+    this.contentMetadata.defaultResource = this.metadataOverrides.defaultResource || this.metadata.defaultResource;
+    this.contentMetadata.streamUrl = this.metadataOverrides.streamUrl || this.metadata.streamUrl;
+
+    return this.contentMetadata;
+  }
+
+  // Those methods should be treated as package private
```
Comment (reevn) [inline src/ts/ContentMetadataBuilder.ts]:
```suggestion
  // These methods should be treated as package private
```

File: src/ts/ConvivaAnalytics.ts (hunk 4 above)
Comment (reevn) [reply src/ts/ConvivaAnalytics.ts]: It's just strange for Austrian brains to see no commas in such a sentence. I think it is valid english syntax though.

Maybe we could do with some commas though?

**Review by reevn — APPROVED** [review]
Code looks fine for me. Can't test the functionality live though 👍

File: src/ts/ConvivaAnalytics.ts (hunk 3 above)
Comment (stonko1994) [reply src/ts/ConvivaAnalytics.ts:137]: Hm ... actually not. `isValidSession` actually don't check if the metadata are valid, only if a session is active.

Not sure if we need to rename it to solve confusion.

File: src/ts/ConvivaAnalytics.ts (hunk 4 above)
Comment (stonko1994) [reply src/ts/ConvivaAnalytics.ts]: What about `// Without setting the asset name on the content metadata the SDK will throw errors when we initialize the session.`

**Review by bitmovin-kenny — APPROVED** [review]
👍

**Pull request merged**